compiler/optimizer/LoopVersioner.cpp

```diff
@@ -8023,6 +8023,9 @@ bool TR_LoopVersioner::requiresPrivatization(TR::Node *node)
    if (!node->getOpCode().hasSymbolReference())
       return false;
 
+   if (node->isDataAddrPointer())
+      return false;
```

Comment:
Most privatizations will be due to `requiresPrivatization()`, i.e. because they appear to be necessary, and this change will prevent them :+1:

However, I'm concerned that versioner might still [heuristically decide to hoist](https://github.com/eclipse/omr/blob/5d2a72e72daeada0b2d205aafb7df5f082a16151/compiler/optimizer/LoopVersioner.cpp#L5737-L5761) a data address load

Could I ask you to try a small OpenJ9 test case and check the log? [DoesItHoist.java](https://github.com/user-attachments/files/15929423/DoesItHoist.java.txt)
```
$ export TR_EnableExpensiveOptsAtWarm=1
$ java -Xjit:'limit=go,count=63,disableAsyncCompilation,initialOptLevel=warm,inhibitRecompilation,disableSuffixLogs,compilationThreads=1,traceFull,log=go.log' DoesItHoist
```
This should privatize the entire load `inst.next.next.next.next.x[32]`, but *not* any internal pointer (`aladd`, data address) needed to perform that load. When I run it without off-heap arrays, it avoids privatizing the `aladd`, as can be seen in the log:
```
n311n     istore  <temp slot 10>[#441  Auto]        <---- privatizing the element load
n310n       iloadi  <array-shadow>[#229  Shadow]
n309n         aladd (internalPtr )                  <---- aladd correctly not privatized
n307n           aload  <temp slot 9>[#440  Auto]    <---- privatized array reference
n308n           lconst 136
```
I think what prevents the privatization here is that `aladd` doesn't satisfy `isSupportedForPRE()`. But `aloadi` does, so maybe the data address will still get privatized. If it does, it looks like it should be possible to prevent that by handling data address pointers in [`addressPointsAtObject()`](https://github.com/eclipse/omr/blob/5d2a72e72daeada0b2d205aafb7df5f082a16151/compiler/il/OMRNode.cpp#L2500-L2515)

I tried to run this with off-heap myself, but I couldn't figure out how to enable it in my build

Reply:
It seem to work correctly, I tested it with a build that has all offheap changes (including https://github.com/eclipse/omr/pull/7368 which has code to disable privatization of the data address too).

```
n365n     BBStart <block_42> (freq 158)                                                       [     0x3fcc59861d0] bci=[-1,20,22] rc=0 vc=0 vn=- li=- udi=- nc=0
n340n     istore  <temp slot 10>[#417  Auto] [flags 0x3 0x0 ]                                 [     0x3fcc5985a00] bci=[-1,46,23] rc=0 vc=0 vn=- li=- udi=- nc=1
n339n       iloadi  <array-shadow>[#254  Shadow] [flags 0x80000603 0x0 ]                      [     0x3fcc59859b0] bci=[-1,46,23] rc=1 vc=0 vn=- li=- udi=- nc=1
n338n         aladd (internalPtr )                                                            [     0x3fcc5985960] bci=[-1,46,23] rc=1 vc=0 vn=- li=- udi=- nc=2 flg=0x8000
n336n           aloadi  <contiguousArrayDataAddrFieldSymbol>[#355  final Shadow +8] [flags 0x20604 0x0 ] (internalPtr )  [     0x3fcc59858c0] bci=[-1,46,23] rc=1 vc=0 vn=- li=- udi=- nc=1 >
n335n             aload  <temp slot 9>[#416  Auto] [flags 0x7 0x0 ]                           [     0x3fcc5985870] bci=[-1,41,23] rc=1 vc=0 vn=- li=- udi=- nc=0
n337n           lconst 128                                                                    [     0x3fcc5985910] bci=[-1,46,23] rc=1 vc=0 vn=- li=- udi=- nc=0
n366n     BBEnd </block_42> =====                                                             [     0x3fcc5986220] bci=[-1,20,22] rc=0 vc=0 vn=- li=- udi=- nc=0
```

Reply:
OK, thanks, I see the change in `isSupportedNodeForFunctionality()` that makes this work

Reply:
I'm OK with it, knowing that's coming in the other PR